### PR TITLE
add quantization backward support

### PR DESCRIPTION
**Pull request by Cjkkkk**

### CODE REVIEW

File: src/sdk/pynni/nni/compression/torch/builtin_quantizers.py [hunk 1]
```diff
@@ -227,20 +227,17 @@ class DoReFaQuantizer(Quantizer):
     (https://arxiv.org/abs/1606.06160)
     """
     def __init__(self, model, config_list):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/torch/builtin_quantizers.py]: better to add docstring to explain `model` and `config_list`

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 2]
```diff
@@ -317,50 +321,78 @@ def _instrument_layer(self, layer, config):
         if 'weight' in config["quant_types"]:
             if not _check_weight(layer.module):
                 _logger.warning('Module %s does not have parameter "weight"', layer.name)
+            else:
+                layer.module.register_parameter("old_weight", torch.nn.Parameter(layer.module.weight))
+                delattr(layer.module, "weight")
+                layer.module.register_buffer('weight', layer.module.old_weight)
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: could add comments here to explain why delete `weight` and then add `weight` again.

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 3]
```diff
@@ -317,50 +321,78 @@ def _instrument_layer(self, layer, config):
         if 'weight' in config["quant_types"]:
             if not _check_weight(layer.module):
                 _logger.warning('Module %s does not have parameter "weight"', layer.name)
+            else:
+                layer.module.register_parameter("old_weight", torch.nn.Parameter(layer.module.weight))
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: can you unify the use of punctuation? such as `"` and `'`

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 4]
```diff
@@ -317,50 +321,78 @@ def _instrument_layer(self, layer, config):
...
-                new_weight = self.quantize_weight(weight, config, op=layer.module, op_type=layer.type, op_name=layer.name)
-                layer.module.weight.data = new_weight
+                new_weight = self.quant_grad.apply(layer.module.old_weight, QuantType.QUANT_WEIGHT, self.quantize_weight, config, layer)
+                layer.module.weight = new_weight
                 result = layer._forward(*inputs)
-                layer.module.weight.data = weight
             else:
                 result = layer._forward(*inputs)
 
             if 'output' in config["quant_types"]:
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: same here.

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 5]
```diff
@@ -317,50 +321,78 @@ def _instrument_layer(self, layer, config):
...
+                layer.module.register_buffer('weight', layer.module.old_weight)
+
         layer._forward = layer.module.forward
 
         def new_forward(*inputs):
             if 'input' in config["quant_types"]:
-                inputs = straight_through_quantize_input.apply(inputs, self, config, layer)
+                inputs = self.quant_grad.apply(inputs, QuantType.QUANT_INPUT, self.quantize_input, config, layer)
 
             if 'weight' in config["quant_types"] and _check_weight(layer.module):
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: same here

File: src/sdk/pynni/nni/compression/torch/builtin_quantizers.py (hunk 1 above)
Comment (Cjkkkk) [reply src/sdk/pynni/nni/compression/torch/builtin_quantizers.py]: fixed

File: src/sdk/pynni/nni/compression/torch/compressor.py (hunk 3 above)
Comment (Cjkkkk) [reply src/sdk/pynni/nni/compression/torch/compressor.py]: fixed

File: src/sdk/pynni/nni/compression/torch/compressor.py (hunk 2 above)
Comment (Cjkkkk) [reply src/sdk/pynni/nni/compression/torch/compressor.py]: fixed

**Pull request merged**